### PR TITLE
fix(engine): pause hardware API when engine is paused

### LINKED ISSUE #7923: refactor(engine): pause HardwareAPI when engine is paused
## Overview

[Note: this ticket is under investigation to determine if this story/spec is desired]

As a JSONv6 protocol operator, when I pause my protocol run, I would like the robot to stop moving as soon as possible.

## Acceptance criteria

- [ ] If a `ProtocolEngine` is paused, motion stops as soon as the hardware control layer is able to stop the motion
    - This may be in the middle of a `ProtocolEngine` command

We will not be implementing a new command or run status string for this change, just the ProtocolEngine <> HardwareAPI interactions when pauses and resumes happen.

## Implementation notes

To implement this behavior, we would add a call to `self._hardware_api.pause(HardwarePauseType.PAUSE)` in `ProtocolEngine.pause` and a call to `self._hardware_api.pause(HardwarePauseType.RESUME)` in `ProtocolEngine.resume`. This will be _in addition to_ the existing action dispatch, which will pause the queue worker, update statuses, etc.

This is exactly what the `LegacyContextPlugin` that translates PAPIv2 runs to PE runs does, so the implementation can more-or-less be cut-and-pasted over. We should take care to avoid interference with #9125.

This nice thing about this change is that while the Hardware API _can_ be paused, it's no longer the mechanism by which the protocol run pause is fundamentally accomplished, so it doesn't violate the architectural reason for ProtocolEngine existing, even though it might feel like it at first glance after years of the HardwareAPI being fundamentally in charge of run pauses

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR adds calls to `hardware_api.pause` and `hardware_api.resume` when `engine.pause` and `engine.resume` are called, respectively.

Closes #7923

## Changelog

- fix(engine): pause hardware API when engine is paused

## Review requests

This PR needs to be tested on actual hardware rather than the simluator or emulator.

We need to check these acceptance criteria with both PAPIv2 protocols _and_ JSONv6 protocols, because some functionality that was scoped specifically to PAPIv2 protocols has now moved into the `ProtocolEngine` proper.

- [ ] When a protocol is paused, it can stop in the middle of a command
    - The command should remain in a `running` state
    - The run should shift to a `paused` state
    - Writing a protocol with big moves can make this easier to test and observe
- [ ] When the protocol is resumed, the command picks up where it left off

### Pause when door is opened

This ticket shouldn't affect "pause on door stop" behavior for PAPIv2 protocols. In advance of #9125, we should check that this PR does not negatively affect our expected PAPIv2 behavior. These tests are not expected to work on JSONv6 protocols until 9125 is completed.

- Before the run is started
    - [ ] Door state does not affect the ability to run LPC (`intent: setup`) commands
- Starting the run
    - [ ] Starting the run (issuing a `play` action) starts the run in a paused state
    - [ ] Trying to resume the run (issuing another `play` action) will reject if the door is still open
    - [ ] If the door is closed _and then_ another `play` action is issued, the run starts
- During the run
    - [ ] If the door is opened, the run becomes paused and the robot stops moving
    - [ ] Trying to resume the run (issuing another `play` action) will reject if the door is still open 
    - [ ] If the door is closed _and then_ another `play` action is issued, the run resumes

## Risk assessment

I started today pretty nervous about this, but once I got past the surface and looked at the ProtocolEngine state behaviors we've already implemented around door state, I'm feeling pretty confident about this!

### CODE REVIEW

**Review by SyntaxColoring — COMMENTED** [review]
I ended up not having a chance to test this on hardware today—sorry!

I've been staring at this for a little bit, but I'm having a pretty hard time following how our play/pause door-opened/door-closed actions flow through all the different layers—runner, engine, plugin, and hardware.

One thing on my mind is this behavior you described in the comment about a run that's played for the first time immediately becoming blocked by the door *before any movement happens.* It's not clear to me how we're implementing that.

Actually, it sort of looks to me like the legacy plugin is competing with `ProtocolEngine.play()`?

1. `ProtocolEngine.play()` will do `self._action_dispatcher.dispatch()`, which will give the legacy plugin a chance to run first (I think?)
2. The plugin will see that the door is blocking, and pause the hardware API
3. `ProtocolEngine.play()` will start the queue worker
4. `ProtocolEngine.play()` will cause `self._hardware_api.resume()`, undoing what the legacy plugin did?

**Comment by mcous** [comment]
I think @SyntaxColoring is right, this PR leaves the `LegacyContextPlugin` fighting with the `ProtocolEngine` for HW pause/resume when the door is opening. Marking as a draft for the time being

File: api/src/opentrons/protocol_engine/protocol_engine.py [hunk 1]
```diff
@@ -107,14 +109,19 @@ def play(self) -> None:
             PlayAction(requested_at=requested_at)
         )
         self._action_dispatcher.dispatch(action)
-        self._queue_worker.start()
+
+        if self._state_store.commands.get_is_door_blocking():
+            self._hardware_api.pause(HardwarePauseType.PAUSE)
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/protocol_engine.py:114]: are there other types of HardwarePauseType we use?

File: api/src/opentrons/protocol_engine/state/config.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+"""Top-level ProtocolEngine configuration options."""
+from dataclasses import dataclass
+
+
+@dataclass(frozen=True)
+class Config:
+    """ProtocolEngine configuration options."""
+
+    ignore_pause: bool = False
+    use_virtual_modules: bool = False
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/state/config.py:19]: is there an explanation somewhere what is the meaning of this variable? I looked in the code and couldn't find

File: api/src/opentrons/protocol_engine/protocol_engine.py (hunk 1 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/protocol_engine.py:114]: The other type is `HardwarePauseType.DELAY`, which is used by legacy code in PAPIv2. The `ProtocolEngine` does not use it

**Comment by mcous** [comment]
I tested 75cea85e on an OT-2 and found the door switch stuff to still be working with PAPIv2. I'm going to mark this as ready for review so pausing the hardware API in the engine natively can happen with #9125

File: api/src/opentrons/hardware_control/pause_manager.py [hunk 3]
```diff
@@ -12,32 +10,18 @@ class PauseManager:
     timer runs out) and the pause resume (trigged by user via the app).
     """
 
-    def __init__(self, door_state: DoorState) -> None:
+    def __init__(self) -> None:
```
Comment (mcous) [inline api/src/opentrons/hardware_control/pause_manager.py:13]: I checked in with @sfoster1 on this one: the HW API would rather simply report door open and close events. This lines up nicely with a desire on our end to stop deeply nesting access to the config singleton.

Given that this PR gives the `ProtocolEngine` enough information to reject an invalid resume while the door is open, I removed the unnecessary door logic from `PauseManager`

File: api/src/opentrons/protocol_engine/state/modules.py [hunk 4]
```diff
@@ -291,7 +291,7 @@ class ModuleView(HasState[ModuleState]):
 
     _state: ModuleState
 
-    def __init__(self, state: ModuleState, virtualize_modules: bool) -> None:
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/modules.py:294]: Argument was completely unused

File: api/src/opentrons/hardware_control/pause_manager.py (hunk 3 above)
Comment (SyntaxColoring) [reply api/src/opentrons/hardware_control/pause_manager.py:13]: My opinion isn't a super informed one, but I'm a big fan of this. It seems way easier to reason about the HW API if we're always controlling it and it's never controlling itself independently.

File: api/src/opentrons/protocol_engine/protocol_engine.py [hunk 5]
```diff
@@ -107,14 +109,19 @@ def play(self) -> None:
             PlayAction(requested_at=requested_at)
         )
         self._action_dispatcher.dispatch(action)
-        self._queue_worker.start()
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/protocol_engine.py:110]: Is this an intentional removal of `self._queue_worker.start()`?

**Review by SyntaxColoring — APPROVED** [review]
Code makes sense to me with the exception of a removal to `queue_worker.start()`. I'll also test this on a robot and report results.

File: robot-server/robot_server/service/session/session_types/live_protocol/session.py [hunk 6]
```diff
@@ -29,7 +34,9 @@ async def create(
         return LiveProtocolSession(
             configuration=configuration,
             instance_meta=instance_meta,
-            protocol_engine=await create_protocol_engine(configuration.hardware),
+            protocol_engine=await create_protocol_engine(
```
Comment (TamarZanzouri) [inline robot-server/robot_server/service/session/session_types/live_protocol/session.py:37]: @mcous you prob notices codecov

File: api/src/opentrons/protocol_engine/protocol_engine.py (hunk 5 above)
Comment (TamarZanzouri) [reply api/src/opentrons/protocol_engine/protocol_engine.py:110]: @SyntaxColoring @mcous  I saw it the tests that instead we are now calling hardware_api.resume(HardwarePauseType.PAUSE)\
 hardware_api.pause(HardwarePauseType.PAUSE)?

**Review by TamarZanzouri — APPROVED** [review]
Changes look good to me and I love the amount of code we got to remove! would love to see what race condition you guys were talking about before you made these changes

File: api/src/opentrons/protocol_engine/protocol_engine.py (hunk 5 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/protocol_engine.py:110]: This is intentional; `queue_worker.start` is called in `__init__` so all subsequent calls will no-op

**Review by sanni-t — APPROVED** [review]
Reviewed in a call w/ @mcous and the changes look great!

**Comment by SyntaxColoring** [comment]
Tested on a robot with this protocol.

```python
metadata = {"apiLevel": "2.12"}

def run(protocol):
    pipette = protocol.load_instrument("p300_multi_gen2", mount="left")
    lw1 = protocol.load_labware("opentrons_96_tiprack_300ul", 1)
    lw2 = protocol.load_labware("opentrons_96_tiprack_300ul", 3)
    for i in range(10):
        protocol.pause()
        for i in range(2):
            pipette.move_to(lw1["A1"].top())
            pipette.move_to(lw2["A12"].top())
```

* The run can still be paused and played, and appears to still handle the door correctly.
* The run state appears to track `running`/`paused`/`blocked-by-door-open` correctly.
* When the run is on a `protocol.pause()` and is waiting for a resume, **the `commands` list (and run log) will show the command *following* the pause command as current.** I'm not sure why.

```json
        {
            "id": "command.PAUSE-1",
            "key": "command.PAUSE-1",
            "commandType": "waitForResume",
            "createdAt": "2022-06-29T16:18:20.701854+00:00",
            "startedAt": "2022-06-29T16:18:20.701854+00:00",
            "completedAt": "2022-06-29T16:18:20.703251+00:00",
            "status": "succeeded",
            "params": {}
        },
        {
            "id": "command.MOVE_TO-4",
            "key": "command.MOVE_TO-4",
            "commandType": "custom",
            "createdAt": "2022-06-29T16:18:20.711125+00:00",
            "startedAt": "2022-06-29T16:18:20.711125+00:00",
            "status": "running",
            "params": {
                "legacyCommandType": "command.MOVE_TO",
                "legacyCommandText": "Moving to A1 of Opentrons 96 Tip Rack 300 µL on 1"
            }
        }
```

**Review by sfoster1 — APPROVED** [review]
Looks great from an embedded point of view - glad to just be providing events and letting someone else worry about it.

**Comment by mcous** [comment]
> When the run is on a protocol.pause() and is waiting for a resume, the commands list (and run log) will show the command following the pause command as current. I'm not sure why.

Expected behavior with PAPIv2, given how pauses work. The `ctx.pause` sets the HW pause flag and returns, allowing the protocol to enter the next `ctx` method (dispatching the message broker event) before the protocol hits a HW call that pauses motion.

This behavior is not the case with JSONv6 protocols, where `waitForResume` will hold in a `running` state until the protocol is resumed

**Pull request merged**